web/modules/custom/lit_cover_service/src/Service/BatchService.php

```diff
@@ -0,0 +1,252 @@
+<?php
+
+namespace Drupal\lit_cover_service\Service;
+
+use Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException;
+use Drupal\Component\Plugin\Exception\PluginNotFoundException;
+use Drupal\Core\Entity\EntityStorageException;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\file\Entity\File;
+use Drupal\node\Entity\Node;
+
+/**
+ * Class BatchService.
+ */
+class BatchService {
+
+  /**
+   * Batch delete process callback.
+   *
+   * @param int $batchId
+   *   Id of the batch
+   * @param $batchTotal
+   *   Total number of batch operations
+   * @param array $nids
+   *   The nids to process
+   * @param object $context
+   *   Context for operations
+   *
+   * @throws \Drupal\Core\Entity\EntityStorageException
+   */
+  public function deleteBookCovers($batchId, $batchTotal, $nids, &$context) {
+    if (!isset($context['sandbox']['progress'])) {
+      $context['sandbox']['progress'] = 0;
+      $context['sandbox']['current_node'] = 0;
+      $context['sandbox']['max'] = $batchTotal;
+    }
+
+    $context['message'] = t('Deleting batch @batchId of @batchTotal', [
+      '@batchId' => $batchId,
+      '@batchTotal' => $batchTotal,
+    ]);
+
+    $nodes = Node::loadMultiple($nids);
+    $deletedImageFiles = [];
+    foreach ($nodes as $node) {
+      $deletedImageFiles[] = self::clearImageField($node);
+      $node->save();
+
+      // Store some result for post-processing in the finished callback.
+      $context['results'][] = $node->nid;
+
+      // Update our progress information.
+      $context['sandbox']['progress']++;
+      $context['sandbox']['current_node'] = $node->nid;
+    }
+
+    self::deleteFileEntities($deletedImageFiles);
+
+    // Inform the batch engine that we are not finished,
+    // and provide an estimation of the completion level we reached.
+    if ($context['sandbox']['progress'] != $context['sandbox']['max']) {
+      $context['finished'] = $context['sandbox']['progress'] / $context['sandbox']['max'];
+    }
+  }
+
+  /**
+   * Batch fetch process callback.
+   *
+   * @param int $batchId
+   *   Id of the batch
+   * @param $batchTotal
+   *   Total number of batch operations
+   * @param array $nids
+   *   The nids to process
+   * @param object $context
+   *   Context for operations
+   *
+   * @throws \Drupal\Core\Entity\EntityStorageException
+   */
+  public function fetchBookCovers($batchId, $batchTotal, $nids, &$context) {
+    if (!isset($context['sandbox']['progress'])) {
+      $context['sandbox']['progress'] = 0;
+      $context['sandbox']['current_node'] = 0;
+      $context['sandbox']['max'] = $batchTotal;
+    }
+
+    $context['message'] = t('Fetching batch @batchId of @batchTotal', [
+      '@batchId' => $batchId,
+      '@batchTotal' => $batchTotal,
+    ]);
+
+    $nodes = Node::loadMultiple($nids);
```

Comment:
If the number of nids is large this could give memory and database problems! On the other hand load each node inside the loop is properly slower?

Reply:
This is true for every batch processor in this file?

Reply:
Number of nodes will never be higher than `BATCH_SIZE` (100) in the command. I considered if the limit should be handled in the command or in the batch service. Decided to stay with the structure suggested in https://www.drupal.org/project/drush9_batch_processing